iroh-gateway/src/handler.rs

```diff
@@ -0,0 +1,308 @@
+                    .await;
+                let body = match body {
+                    Ok(b) => b,
+                    Err(e) => {
+                        let msg = format!("{}", e);
+                        return Handler::error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+                    }
+                };
+
+                headers = Handler::set_content_disposition_headers(
+                    &headers,
+                    format!("{}.bin", cid).as_str(),
+                    "attachment",
+                );
+                Handler::response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+            }
+            ResponseFormat::Car => {
+                let body = client
+                    .get_file(format!("{}", full_content_path).as_str())
+                    .await;
+                let body = match body {
+                    Ok(b) => b,
+                    Err(e) => {
+                        let msg = format!("{}", e);
+                        return Handler::error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+                    }
+                };
+                headers = Handler::set_content_disposition_headers(
+                    &headers,
+                    format!("{}.car", cid).as_str(),
+                    "attachment",
+                );
+                Handler::response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+            }
+            ResponseFormat::HTML => {
+                let body = format!("<p>{}</p>", cid);
+                Handler::response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+            }
+            ResponseFormat::FS => {
+                let body = client
+                    .get_file(format!("{}", full_content_path).as_str())
+                    .await;
+                let body = match body {
+                    Ok(b) => b,
+                    Err(e) => {
+                        let msg = format!("{}", e);
+                        return Handler::error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+                    }
+                };
+                let (name, headers) = Handler::add_content_disposition_headers(
+                    &headers,
+                    &query_file_name,
+                    &cpath,
+                    download,
+                );
+                let headers = Handler::add_content_type_headers(&headers, &name);
+                Handler::response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+            }
+        }
+    }
+
+    fn response_format(format: &str) -> Result<ResponseFormat, String> {
+        match format {
+            "raw" => Ok(ResponseFormat::Raw),
+            "car" => Ok(ResponseFormat::Car),
+            "html" => Ok(ResponseFormat::HTML),
+            "" => Ok(ResponseFormat::FS),
+            _ => Err("format not supported".to_string()),
+        }
+    }
+
+    fn format_headers(format: &ResponseFormat) -> HashMap<String, String> {
```

Comment:
all of the methods below should take a mutable reference to a hashmap and insert into that, instead lf creating new hashmaps

Reply:
yep way too much cloning going around :D